drivers/hmis_csv_importer/app/models/hmis_csv_importer/importer/importer.rb

```diff
@@ -101,22 +108,224 @@ def resume!
+    # are aggregated correctly without raising additional errors if files are missing.
+    #
+    # @return [Hash{String => Integer}] A hash where keys are filenames and values are the total error counts for each file.
+    #
+    memoize private def error_counts
+      file_lookup = self.class.importable_files_map.invert
+
+      summary = @loader_log.summary
+      # Initialize error counts grouped by filename
+      counts = summary.map { |filename, data| [filename, data['total_errors'].to_i] }.to_h
+
+      # Collect errors from ImportError model
+      HmisCsvImporter::Importer::ImportError.where(
         importer_log_id: importer_log.id,
-      )
+      ).
+        group(:source_type).
+        count.
+        map do |k, count|
+          # convert to filename keys
+          filename = file_lookup[k.demodulize]
+          # every file _should_ be present, but let's try not to throw any errors here
+          next unless counts[filename]
+
+          counts[filename] += count
+        end
 
-      validation_errors = HmisCsvImporter::HmisCsvValidation::Base.where(
+      # Collect validation errors
+      HmisCsvImporter::HmisCsvValidation::Base.where(
         type: HmisCsvImporter::HmisCsvValidation::Error.subclasses.map(&:name),
         importer_log_id: importer_log.id,
-      )
+      ).
+        group(:source_type).
+        count.
+        map do |k, count|
+          # convert to filename keys
+          filename = file_lookup[k.demodulize]
+          # every file _should_ be present, but let's try not to throw any errors here
+          next unless counts[filename]
+
+          counts[filename] += count
+        end
+      counts
+    end
+
+    ##
+    # Estimates the number of records that will be added and removed during the import process.
+    #
+    # This method calculates changes before ingestion to determine if the import
+    # exceeds configured thresholds. It compares existing records in the warehouse
+    # with incoming records from the import log to identify additions and removals.
+    #
+    # The results are stored in the `importer_log.summary` for each file, allowing these values
+    # to be used in import threshold checks before committing data changes.
+    #
+    # @return [void]
+    #
+    private def precalculate_change_counts
+      # This makes an estimate of the changes that will occur as it needs to be done before the
+      # actual processing so that it can pause the import if necessary.
+      # NOTE: as written this causes 2 additional, potentially slow queries
+      importable_files.map do |file, klass|
+        incoming_data = klass.incoming_data(importer_log_id: importer_log.id).
+          pluck(klass.hud_key).to_set
+
+        to_add = (incoming_data - existing_hud_keys(klass)).count
+        to_remove = (existing_hud_keys(klass) - incoming_data).count
+        # If we never delete, just pretend it will be 0
+        to_remove = 0 if klass.prevent_import_deletions?
```

Comment:
conditional assignment to avoid unnecessary `to_remove` calculation
```suggestion
        # If we never delete, just pretend it will be 0
        to_remove = 0 if klass.prevent_import_deletions?
        to_remove ||= (existing_hud_keys(klass) - incoming_data).count

```

Reply:
Thanks!